File: sdk/js/src/token.ts
/**
 * Token minting and verification for Agent-Safe capability tokens.
 * Uses Ed25519 for signing via node:crypto (zero deps).
 */

import { createPrivateKey, createPublicKey, sign, generateKeyPairSync, createHash } from 'node:crypto';
import { parseSExpr, evalPolicy } from './spl.js';
import { verifyEd25519, sha256 } from './crypto.js';

export interface Token {
  version: string;
  policy: string;
  merkle_root?: string;
  hash_chain_commitment?: string;
  sealed: boolean;
  expires?: string;
  public_key: string;
  signature: string;
  pop_key?: string;
}

export interface MintOptions {
  merkleRoot?: string;
  hashChainCommitment?: string;
  sealed?: boolean;
  expires?: string;
  popKey?: string;
}

export interface VerifyTokenOptions {
  vars?: Record<string, any>;
  per_day_count?: (action: string, day: string) => number;
  crypto?: {
    dpop_ok?: () => boolean;
    merkle_ok?: (tuple: any[]) => boolean;
    vrf_ok?: (day: string, amount: number) => boolean;
    thresh_ok?: () => boolean;
  };
  now?: string;
  presentationSignature?: string;
}

/**
 * Generate an Ed25519 keypair.
 * @returns { publicKey: hex, privateKey: hex }
 */
export function generateKeypair(): { publicKey: string; privateKey: string } {
  const { publicKey, privateKey } = generateKeyPairSync('ed25519');
  const pubRaw = publicKey.export({ type: 'spki', format: 'der' }).subarray(12); // strip DER prefix
  const privRaw = privateKey.export({ type: 'pkcs8', format: 'der' }).subarray(16); // strip DER prefix
  return {
    publicKey: Buffer.from(pubRaw).toString('hex'),
    privateKey: Buffer.from(privRaw).toString('hex'),
  };
}

/**
 * Build the canonical signing payload for a token.
 * Covers all security-relevant fields so sealed, expires, merkle_root, and
 * hash_chain_commitment cannot be tampered with after signing.
 */
export function signingPayload(
  policy: string,
  merkleRoot?: string,
  hashChainCommitment?: string,
  sealed?: boolean,
  expires?: string,
): Buffer {
  const parts = [
    policy.trim(),
    merkleRoot ?? '',
    hashChainCommitment ?? '',
    sealed ? '1' : '0',
    expires ?? '',
  ];
  return Buffer.from(parts.join('\0'), 'utf8');
}

/**
 * Create a PoP presentation signature for a token.
 * The agent signs SHA-256(signing_payload) with its own Ed25519 key.
 * @param token - Token to present
 * @param agentPrivateKeyHex - Agent's Ed25519 private key (32 bytes seed, hex)
 * @returns Hex-encoded presentation signature
 */
export function createPresentationSignature(token: Token, agentPrivateKeyHex: string): string {
  const payload = signingPayload(
    token.policy, token.merkle_root, token.hash_chain_commitment, token.sealed, token.expires,
  );
  const popPayload = sha256(payload);

  const privSeed = Buffer.from(agentPrivateKeyHex, 'hex');
  const pkcs8Prefix = Buffer.from('302e020100300506032b657004220420', 'hex');
  const privDer = Buffer.concat([pkcs8Prefix, privSeed]);
  const keyObj = createPrivateKey({ key: privDer, format: 'der', type: 'pkcs8' });

  const sig = sign(null, Buffer.from(popPayload), keyObj);
  return Buffer.from(sig).toString('hex');
}

/**
 * Mint a signed capability token.
 * @param policy - SPL policy source string
 * @param privateKeyHex - Hex-encoded Ed25519 private key (32 bytes seed)
 * @param options - Optional token metadata
 * @returns Signed token object
 */
export function mint(policy: string, privateKeyHex: string, options: MintOptions = {}): Token {
  const sealedVal = options.sealed ?? false;
  const payload = signingPayload(
    policy, options.merkleRoot, options.hashChainCommitment, sealedVal, options.expires,
  );

  // Reconstruct DER-encoded PKCS8 private key
  const privSeed = Buffer.from(privateKeyHex, 'hex');
  const pkcs8Prefix = Buffer.from('302e020100300506032b657004220420', 'hex');
  const privDer = Buffer.concat([pkcs8Prefix, privSeed]);

  const keyObj = createPrivateKey({ key: privDer, format: 'der', type: 'pkcs8' });
  const signature = sign(null, payload, keyObj);

  // Derive public key
  const pubKey = createPublicKey(keyObj);
  const pubDer = pubKey.export({ type: 'spki', format: 'der' });
  const pubRaw = pubDer.subarray(12);

  const token: Token = {
    version: '0.1.0',
    policy: policy.trim(),
    merkle_root: options.merkleRoot,
    hash_chain_commitment: options.hashChainCommitment,
    sealed: sealedVal,
    expires: options.expires,
    public_key: Buffer.from(pubRaw).toString('hex'),
    signature: Buffer.from(signature).toString('hex'),
  };
  if (options.popKey) token.pop_key = options.popKey;
  return token;
}

/**
 * Verify a token and evaluate its policy against a request.
 * @param token - Token object or JSON string
 * @param req - Request object
 * @param options - Verification options (vars, crypto callbacks, etc.)
 * @returns { allow: boolean, token: Token }
 */
export function verifyToken(
  token: Token | string,
  req: Record<string, any>,
  options: VerifyTokenOptions = {},
): { allow: boolean; sealed: boolean; error?: string } {
  const t: Token = typeof token === 'string' ? JSON.parse(token) : token;

  // Check expiration
  if (t.expires && new Date(t.expires) < new Date(options.now ?? new Date().toISOString())) {
    return { allow: false, sealed: t.sealed, error: 'token expired' };
  }

  // Verify signature over full token envelope
  const payload = signingPayload(
    t.policy, t.merkle_root, t.hash_chain_commitment, t.sealed, t.expires,
  );
  if (!verifyEd25519(payload, t.signature, t.public_key)) {
    return { allow: false, sealed: t.sealed, error: 'invalid signature' };
  }

  // PoP binding: if token has pop_key, require and verify presentation signature
  if (t.pop_key) {
    if (!options.presentationSignature) {
      return { allow: false, sealed: t.sealed, error: 'PoP binding requires presentation signature' };
    }
    const popPayload = sha256(payload);
    if (!verifyEd25519(popPayload, options.presentationSignature, t.pop_key)) {
      return { allow: false, sealed: t.sealed, error: 'invalid presentation signature' };
    }
  }

  // If sealed, token is valid but cannot be attenuated — evaluate normally
  // (Sealing prevents attenuation, not verification)

  // Parse and evaluate
  const ast = parseSExpr(t.policy);
  const ctx = {
    req,
    vars: options.vars ?? {},
    now: options.now ?? new Date().toISOString(),
    per_day_count: options.per_day_count ?? (() => 0),
    crypto: {
      dpop_ok: options.crypto?.dpop_ok ?? (() => false),
      merkle_ok: options.crypto?.merkle_ok ?? (() => false),
      vrf_ok: options.crypto?.vrf_ok ?? (() => false),
      thresh_ok: options.crypto?.thresh_ok ?? (() => false),
    },
  };
  const allow = !!evalPolicy(ast, ctx);

  return { allow, sealed: t.sealed };
}

File: sdk/rust/src/token.rs
use ed25519_dalek::{SigningKey, Signer};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::HashMap;

use crate::crypto::verify_ed25519;
use crate::evaluator::eval_policy;
use crate::parser::parse;
use crate::types::{CryptoCallbacks, Env, Node, SplError};

/// A signed Agent-Safe capability token.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Token {
    pub version: String,
    pub policy: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub merkle_root: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hash_chain_commitment: Option<String>,
    pub sealed: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires: Option<String>,
    pub public_key: String,
    pub signature: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pop_key: Option<String>,
}

/// Options for minting a token.
#[derive(Default)]
pub struct MintOptions {
    pub merkle_root: Option<String>,
    pub hash_chain_commitment: Option<String>,
    pub sealed: bool,
    pub expires: Option<String>,
    pub pop_key: Option<String>,
}

/// Generate an Ed25519 keypair.
/// Returns (public_key_hex, private_key_hex).
pub fn generate_keypair() -> (String, String) {
    let mut seed = [0u8; 32];
    getrandom::fill(&mut seed).expect("OS RNG failed");
    let signing_key = SigningKey::from_bytes(&seed);
    let verifying_key = signing_key.verifying_key();
    (
        hex::encode(verifying_key.as_bytes()),
        hex::encode(signing_key.as_bytes()),
    )
}

/// Build the canonical signing payload for a token.
/// Covers all security-relevant fields so sealed, expires, merkle_root, and
/// hash_chain_commitment cannot be tampered with after signing.
pub fn signing_payload(
    policy: &str,
    merkle_root: &Option<String>,
    hash_chain_commitment: &Option<String>,
    sealed: bool,
    expires: &Option<String>,
) -> Vec<u8> {
    let parts = [
        policy.trim(),
        merkle_root.as_deref().unwrap_or(""),
        hash_chain_commitment.as_deref().unwrap_or(""),
        if sealed { "1" } else { "0" },
        expires.as_deref().unwrap_or(""),
    ];
    parts.join("\0").into_bytes()
}

/// Mint a signed capability token.
pub fn mint(policy: &str, private_key_hex: &str, opts: MintOptions) -> Result<Token, SplError> {
    let seed_bytes = hex::decode(private_key_hex)
        .map_err(|e| SplError(format!("invalid private key hex: {e}")))?;
    let seed: [u8; 32] = seed_bytes
        .try_into()
        .map_err(|_| SplError("private key must be 32 bytes".to_string()))?;

    let signing_key = SigningKey::from_bytes(&seed);
    let verifying_key = signing_key.verifying_key();

    let payload = signing_payload(
        policy, &opts.merkle_root, &opts.hash_chain_commitment, opts.sealed, &opts.expires,
    );
    let signature = signing_key.sign(&payload);

    Ok(Token {
        version: "0.1.0".to_string(),
        policy: policy.trim().to_string(),
        merkle_root: opts.merkle_root,
        hash_chain_commitment: opts.hash_chain_commitment,
        sealed: opts.sealed,
        expires: opts.expires,
        public_key: hex::encode(verifying_key.as_bytes()),
        signature: hex::encode(signature.to_bytes()),
        pop_key: opts.pop_key,
    })
}

/// Create a PoP presentation signature for a token.
/// The agent signs SHA-256(signing_payload) with its own Ed25519 key.
pub fn create_presentation_signature(
    token: &Token,
    agent_private_key_hex: &str,
) -> Result<String, SplError> {
    let seed_bytes = hex::decode(agent_private_key_hex)
        .map_err(|e| SplError(format!("invalid agent private key hex: {e}")))?;
    let seed: [u8; 32] = seed_bytes
        .try_into()
        .map_err(|_| SplError("agent private key must be 32 bytes".to_string()))?;

    let signing_key = SigningKey::from_bytes(&seed);
    let payload = signing_payload(
        &token.policy, &token.merkle_root, &token.hash_chain_commitment,
        token.sealed, &token.expires,
    );
    let mut hasher = Sha256::new();
    hasher.update(&payload);
    let pop_payload = hasher.finalize();

    let sig = signing_key.sign(&pop_payload);
    Ok(hex::encode(sig.to_bytes()))
}

/// Result of token verification.
pub struct VerifyTokenResult {
    pub allow: bool,
    pub sealed: bool,
    pub error: Option<String>,
}

/// Verify a token's signature and evaluate its policy.
pub fn verify_token(
    token: &Token,
    req: HashMap<String, Node>,
    vars: HashMap<String, Node>,
) -> VerifyTokenResult {
    verify_token_with_pop(token, req, vars, None)
}

/// Verify a token with optional PoP presentation signature.
pub fn verify_token_with_pop(
    token: &Token,
    req: HashMap<String, Node>,
    vars: HashMap<String, Node>,
    presentation_signature: Option<&str>,
) -> VerifyTokenResult {
    // Verify signature over full token envelope
    let payload = signing_payload(
        &token.policy, &token.merkle_root, &token.hash_chain_commitment,
        token.sealed, &token.expires,
    );
    if !verify_ed25519(
        &payload,
        &token.signature,
        &token.public_key,
    ) {
        return VerifyTokenResult {
            allow: false,
            sealed: token.sealed,
            error: Some("invalid signature".to_string()),
        };
    }

    // PoP binding: if token has pop_key, require and verify presentation signature
    if let Some(pop_key) = &token.pop_key {
        match presentation_signature {
            None => {
                return VerifyTokenResult {
                    allow: false,
                    sealed: token.sealed,
                    error: Some("PoP binding requires presentation signature".to_string()),
                };
            }
            Some(pres_sig) => {
                let mut hasher = Sha256::new();
                hasher.update(&payload);
                let pop_payload = hasher.finalize();
                if !verify_ed25519(&pop_payload, pres_sig, pop_key) {
                    return VerifyTokenResult {
                        allow: false,
                        sealed: token.sealed,
                        error: Some("invalid presentation signature".to_string()),
                    };
                }
            }
        }
    }

    // Parse policy
    let ast = match parse(&token.policy) {
        Ok(ast) => ast,
        Err(e) => {
            return VerifyTokenResult {
                allow: false,
                sealed: token.sealed,
                error: Some(format!("parse error: {e}")),
            }
        }
    };

    // Evaluate
    let env = Env {
        req,
        vars,
        per_day_count: Box::new(|_, _| 0),
        crypto: CryptoCallbacks::default(),
        max_gas: 10_000,
        sealed: false,
        strict: false,
    };

    match eval_policy(&ast, &env) {
        Ok(result) => VerifyTokenResult {
            allow: result.is_truthy(),
            sealed: token.sealed,
            error: None,
        },
        Err(e) => VerifyTokenResult {
            allow: false,
            sealed: token.sealed,
            error: Some(e.to_string()),
        },
    }
}
